File: 10_04_2023/q26_Remove_Duplicates_from_Sorted_Array.py
import collections

"""
    Questions: 
"""

class Solution:
    def remove_duplicates(self, nums: list) -> int:
        """
            Time & space complexity analysis.
        """
        print(nums)

        i = 1
        while(i < len(nums)):

            if nums[i] == nums[i-1]:
                print(f'Pop up {i} - {nums[i]}')
                nums.pop(i)
                print(nums)
                i -= 1

            i += 1

        return len(nums)


if __name__ == '__main__':
    # input examples
    # nums = [1, 1, 2]
    nums = [0,0,1,1,1,2,2,3,3,4]
    print(Solution().remove_duplicates(nums))

File: 10_18_2023-26_11_134/Q26RemoveDuplicatesFromSortedArray.java
class Q26RemoveDuplicatesFromSortedArray {

    public int removeDuplicates(int[] nums) {
        int len = nums.length;
        int i = 1;
        int j = 1;

        // 1) if i==0 and j==0 -> i++; j++;

        // 2) if i==j:
        //          if nums[i-1] != nums[i] -> i++; j++;
        //          else j++;
        //    else: #i!=j
        //          if nums[j] == nums[i-1] -> j++;
        //          else nums[i] = nums[j] -> i++; j++

        while(j < len){
            if (i==j) {
                if (nums[i-1] != nums[i]) {
                    i++;
                }
            } else {
                if (nums[i-1] != nums[j]) {
                    nums[i] = nums[j];
                    i++;
                }
            }
            j++;
        }

        return i;
    }

    public static void main(String[] args){
        int[] nums = {1, 1, 2};
        Q26RemoveDuplicatesFromSortedArray sol = new Q26RemoveDuplicatesFromSortedArray();
        System.out.println(sol.removeDuplicates(nums));
    }

}